File: cpp/problem025.cpp
#include <iostream>
#include <cmath>


using namespace std;



int main() {
    // using fibonacci estimation equation: round(phi^n / sqrt(5))
    // floor(log(phi^n / sqrt(5)) + 1) = floor(log(phi^n) - log(sqrt(5))) + 1 =
    // = floor(n * log(phi) - log(sqrt(5))) + 1
    double n = 1000.0, k = 1.0;
    double r5 = sqrt(5.0);
    double phi = (r5 + 1.0) / 2.0;
    double lp = log10(phi);
    double lr5 = log10(r5);
    while (k * lp - lr5 + 1.0 < n) k++;
    cout << k << endl;
    return 0;    
}

File: python/problem025.py
import math

def bruteforce(ndigits):
    a, b, i = 1, 1, 2
    while len(str(b)) < ndigits:
        i += 1
        c = a + b
        a = b
        b = c
    return b, i

def estimate(ndigits):
    phi = (1.0 + math.sqrt(5.0)) / 2.0
    lp = math.log(phi, 10)
    k = 1
    while math.floor(lp * k - math.log(math.sqrt(5), 10)) + 1 < ndigits:
        k += 1

    return math.ceil(lp * k) + 1, k

print(bruteforce(1000))
print(estimate(1000))
    
